src/charm.py

```diff
@@ -2,22 +2,208 @@
+        )
+        return content
+
+    def _push_config_file(
+        self,
+        content: str,
+    ) -> None:
+        """Writes the NSSF config file and pushes it to the container.
+
+        Args:
+            content (str): Content of the config file.
+        """
+        self._nssf_container.push(
+            path=f"{CONFIG_DIR}/{CONFIG_FILE_NAME}",
+            source=content,
+            make_dirs=True,
+        )
+        logger.info("Pushed %s config file", CONFIG_FILE_NAME)
+
+    def _relation_created(self, relation_name: str) -> bool:
+        """Returns True if the relation is created, False otherwise.
+
+        Args:
+            relation_name (str): Name of the relation.
+
+        Returns:
+            bool: True if the relation is created, False otherwise.
+        """
+        return bool(self.model.get_relation(relation_name))
+
+    @property
+    def _nssf_pebble_layer(self) -> Layer:
+        """Returns pebble layer for the nssf container.
+
+        Returns:
+            Layer: Pebble Layer
+        """
+        return Layer(
+            {
+                "services": {
+                    self._nssf_service_name: {
+                        "override": "replace",
+                        "startup": "enabled",
+                        "command": f"/free5gc/nssf/nssf --nssfcfg {CONFIG_DIR}/{CONFIG_FILE_NAME}",  # noqa: E501
+                        "environment": self._nssf_environment_variables,
+                    },
+                },
+            }
+        )
+
+    @property
+    def _nssf_environment_variables(self) -> dict:
+        """Returns environment variables for the nssf container.
+
+        Returns:
+            dict: Environment variables.
+        """
+        return {
+            "GOTRACEBACK": "crash",
+            "GRPC_GO_LOG_VERBOSITY_LEVEL": "99",
+            "GRPC_GO_LOG_SEVERITY_LEVEL": "info",
+            "GRPC_TRACE": "all",
+            "GRPC_VERBOSITY": "DEBUG",
+            "POD_IP": self._pod_ip,
+            "MANAGED_BY_CONFIG_POD": "true",
+        }
+
+    @property
+    def _pod_ip(
+        self,
+    ) -> str:
+        """Returns the pod IP using juju client.
+
+        Returns:
+            str: The pod IP.
+        """
+        return str(IPv4Address(check_output(["unit-get", "private-address"]).decode().strip()))
+
+    @property
+    def _nrf_data_is_available(self) -> bool:
+        """Returns whether the NRF data is available.
+
+        Returns:
+            bool: Whether the NRF data is available.
+        """
+        if not self._nrf_requires.nrf_url:
+            return False
+        return True
+
+    @property
+    def _nssf_hostname(self) -> str:
+        """Builds and returns the NSSF hostname in the cluster.
+
+        Returns:
+            str: The NSSF hostname.
+        """
+        return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
 
 
-if __name__ == "__main__":  # pragma: nocover
+if __name__ == "__main__":  # pragma: no cover
```

Comment:
Why do we need this ignore? I've never seen it before.

Reply:
Exactly, this is just telling the code coverage to ignore this block, as we will never test it in unit tests.